File: rust-client/examples/deposit_transfer_withdraw.rs
use anyhow::{anyhow, Result};
use rust_client_example::{cli_keypair, landed_slot, setup, SetupContext};
use solana_keypair::Keypair;
use solana_signer::Signer;
use zolana_client::{IndexerRpcConfig, Rpc, SolanaRpc, ZolanaClient};
use zolana_interface::instruction::{
    AssetDeposit, Deposit, DepositAsset, Transact, TransactInterfaceTransferAccounts,
    TransactSolTransferAccounts,
};
use zolana_keypair::ShieldedKeypair;
use zolana_transaction::{
    decrypt_spendable, instructions::transact::ConfidentialTransaction, AssetRegistry, SOL_MINT,
};

const DEPOSIT_AMOUNT: u64 = 10_000_000;
const TRANSFER_AMOUNT: u64 = 3_000_000;
const WITHDRAW_AMOUNT: u64 = 3_000_000;

fn main() -> Result<()> {
    let SetupContext {
        rpc_url,
        indexer_url,
        prover_url,
        tree,
    } = setup()?;

    // Connect to the RPC, indexer, and prover.
    let client = ZolanaClient::from_urls(SolanaRpc::new(rpc_url), &indexer_url, prover_url)?;

    // Mints that are registered with Solana Rings for privacy.
    let assets = AssetRegistry::default();
    // SPL: assets.insert(spl.asset_id, spl.mint)?;

    // Initialize the sender's private wallet and local authority
    // to decrypt transactions and sync balances.
    // The Solana signer and private wallet are derived from the same Ed25519 seed.
    let sender = ShieldedKeypair::from_keypair(&cli_keypair()?)?;
    let recipient = ShieldedKeypair::from_keypair(&Keypair::new())?;
    let sender_shielded_address = sender.shielded_address()?;

    // Deposit SOL into the sender's private balance.
    // A deposit from a public balance reveals
    // sender, recipient, asset and amount.
    // Alternatively, you can onramp fiat directly to a private balance.

    // 1. Move public SOL into the sender's private balance.
    let sender_balances_after_deposit = {
        let deposit_ix = Deposit {
            tree,
            depositor: sender.pubkey(),
            deposits: vec![AssetDeposit {
                asset: DepositAsset::Sol,
                // SPL: asset: DepositAsset::Spl(zolana_interface::instruction::DepositSplAccounts {
                // SPL:     mint: spl.mint,
                // SPL:     user_token: spl.user_token_account,
                // SPL:     token_program: spl.token_program,
                // SPL: }),
                view_tag: sender_shielded_address.confidential_view_tag()?,
                owner: sender_shielded_address.owner_hash()?,
                amount: DEPOSIT_AMOUNT,
                utxo_data: None,
                memo: None,
            }],
        }
        .instruction()?;

        // 2. Send and confirm like any Solana transaction; the landed slot gates
        // the indexer fetch below.
        let signature = client.create_and_send_transaction(
            &[deposit_ix],
            sender.pubkey(),
            &[&sender],
            client.compute_budget(),
        )?;
        let slot = landed_slot(&client, signature)?;

        // 3. Fetch this transaction's outputs, gated on its confirmed slot.
        let response = client.get_shielded_transactions_by_signature(
            signature,
            Some(IndexerRpcConfig::at_slot(slot)),
        )?;
        let transactions = response
            .transactions
            .into_iter()
            .map(|indexed| indexed.transaction)
            .collect::<Vec<_>>();

        // 4. The sender decrypts the transaction outputs locally to read the funds deposited in this run.
        let balances = decrypt_spendable(&sender, &transactions, &assets)
            .map_err(|e| anyhow!("decrypt sender transactions: {e:?}"))?
            .balances;

        let sender_balance = balances
            .get_balance(SOL_MINT)
            // SPL: .get_balance(spl.mint)
            .expect("failed to fetch sender's utxo");
        assert_eq!(sender_balance.amount, DEPOSIT_AMOUNT);
        assert_eq!(sender_balance.utxos.len(), 1);

        balances
    };

    // Confidential SOL transfer to the recipient's private balance.
    // A confidential transfer reveals only sender and recipient,
    // not the asset or amount.
    let sender_balances_after_transfer = {
        // 1. Select UTXOs that make up the private balance for the transfer.
        let transfer_utxo = sender_balances_after_deposit
            .get_balance(SOL_MINT)
            // SPL: .get_balance(spl.mint)
            .and_then(|balance| balance.utxos.first())
            .expect("failed to fetch deposited utxo")
            .clone();

        // 2. Prepare the selected UTXOs as inputs for the zero-knowledge proof.
        let mut transfer = ConfidentialTransaction::new(vec![transfer_utxo], sender.pubkey())?;

        // 3. Build and encrypt the confidential transfer.
        // Encryption hides the asset and amount and produces the proof inputs for the ZK prover.
        transfer.transfer_sol(&recipient.shielded_address()?, TRANSFER_AMOUNT)?;
        // SPL: transfer.transfer(&recipient.shielded_address()?, spl.mint, TRANSFER_AMOUNT)?;
        let proof_inputs = transfer.encrypt(&sender)?;

        // 4. Fetch the zk proof to prove the sender can spend the balance without revealing asset and amount.
        let transfer_data = client.prove_transact(proof_inputs, None, &sender)?;

        // 5. Construct the instruction.
        let transfer_ix = Transact {
            payer: sender.pubkey(),
            input_trees: vec![tree],
            output_tree: tree,
            owner_signers: Vec::new(),
            interface_transfer_accounts: Vec::new(),
            data: transfer_data,
        }
        .instruction();

        // 6. Send and confirm like any Solana transaction; confirmation yields the landed slot.
        let signature = client.create_and_send_transaction(
            &[transfer_ix],
            sender.pubkey(),
            &[&sender],
            client.compute_budget(),
        )?;
        let slot = landed_slot(&client, signature)?;

        // 7. Fetch this transaction's outputs, gated on its confirmed slot.
        let response = client.get_shielded_transactions_by_signature(
            signature,
            Some(IndexerRpcConfig::at_slot(slot)),
        )?;
        let transactions = response
            .transactions
            .into_iter()
            .map(|indexed| indexed.transaction)
            .collect::<Vec<_>>();
        let sender_balances = decrypt_spendable(&sender, &transactions, &assets)
            .map_err(|e| anyhow!("decrypt sender transactions: {e:?}"))?
            .balances;
        let sender_balance = sender_balances
            .get_balance(SOL_MINT)
            // SPL: .get_balance(spl.mint)
            .expect("failed to fetch sender's utxo");
        assert_eq!(sender_balance.amount, DEPOSIT_AMOUNT - TRANSFER_AMOUNT);
        assert_eq!(sender_balance.utxos.len(), 1);

        sender_balances
    };

    // Withdraw SOL back to the sender's public balance.
    // A withdrawal from a confidential balance reveals
    // sender, recipient, asset and amount.
    {
        // 1. Select UTXOs that make up the private balance for the withdrawal.
        let withdrawal_utxo = sender_balances_after_transfer
            .get_balance(SOL_MINT)
            // SPL: .get_balance(spl.mint)
            .and_then(|balance| balance.utxos.first())
            .expect("failed to fetch sender's utxo")
            .clone();

        // 2. Prepare the selected UTXOs as inputs for the zero-knowledge proof.
        let mut withdrawal = ConfidentialTransaction::new(vec![withdrawal_utxo], sender.pubkey())?;

        // 3. Build and encrypt the confidential withdrawal.
        // Encryption hides the private change and produces the ZK prover inputs.
        withdrawal.withdraw_sol(WITHDRAW_AMOUNT, sender.pubkey())?;
        // SPL: withdrawal.withdraw(spl.mint, WITHDRAW_AMOUNT, spl.user_token_account)?;
        let proof_inputs = withdrawal.encrypt(&sender)?;

        // 4. Fetch the ZK proof to prove the sender can spend the balance.
        let withdrawal_data = client.prove_transact(proof_inputs, None, &sender)?;

        // 5. Combine the proof and withdrawal accounts in a single instruction.
        let withdraw_ix = Transact {
            payer: sender.pubkey(),
            input_trees: vec![tree],
            output_tree: tree,
            owner_signers: Vec::new(),
            interface_transfer_accounts: vec![TransactInterfaceTransferAccounts::Sol(
                TransactSolTransferAccounts {
                    recipient: sender.pubkey(),
                },
            )],
            // SPL: interface_transfer_accounts: vec![
            // SPL:     TransactInterfaceTransferAccounts::SplWithdrawal(
            // SPL:         zolana_interface::instruction::TransactSplWithdrawalAccounts {
            // SPL:             mint: spl.mint,
            // SPL:             vault: spl.vault,
            // SPL:             user_token_account: spl.user_token_account,
            // SPL:             token_program: spl.token_program,
            // SPL:         },
            // SPL:     ),
            // SPL: ],
            data: withdrawal_data,
        }
        .instruction();

        // 6. Send and confirm like any Solana transaction.
        let signature = client.create_and_send_transaction(
            &[withdraw_ix],
            sender.pubkey(),
            &[&sender],
            client.compute_budget(),
        )?;
        let slot = landed_slot(&client, signature)?;

        // 7. Fetch this transaction's outputs, gated on its confirmed slot.
        let response = client.get_shielded_transactions_by_signature(
            signature,
            Some(IndexerRpcConfig::at_slot(slot)),
        )?;
        let transactions = response
            .transactions
            .into_iter()
            .map(|indexed| indexed.transaction)
            .collect::<Vec<_>>();
        let sender_balances = decrypt_spendable(&sender, &transactions, &assets)
            .map_err(|e| anyhow!("decrypt sender transactions: {e:?}"))?
            .balances;
        let sender_balance = sender_balances
            .get_balance(SOL_MINT)
            // SPL: .get_balance(spl.mint)
            .expect("failed to fetch sender's utxo");
        assert_eq!(
            sender_balance.amount,
            DEPOSIT_AMOUNT - TRANSFER_AMOUNT - WITHDRAW_AMOUNT
        );
        assert_eq!(sender_balance.utxos.len(), 1);

        // 8. Read the funds remaining from this run and the public SOL balance.
        let solana_balance = client.get_balance(sender.pubkey())?;
        println!("withdraw solana_balance={solana_balance} tx={signature}");
        // SPL: println!(
        // SPL:     "withdraw user_token={} tx={signature}",
        // SPL:     spl.user_token_account,
        // SPL: );
    }
    Ok(())
}

File: typescript-client/examples/deposit_transfer_withdraw.ts
import {
  SOL_MINT,
  createZolanaClient,
} from "@heliuslabs/zolana";
import { atSlot } from "@heliuslabs/zolana/client";
import {
  depositInstruction,
  transactInstruction,
  DepositAsset,
  TransactWithdrawal,
} from "@heliuslabs/zolana/interface";
import { randomBlinding } from "@heliuslabs/zolana/keypair";
import {
  AssetRegistry,
  ConfidentialTransfer,
  ProofInputUtxo,
  decryptToBalances,
  WithdrawalTarget,
} from "@heliuslabs/zolana/transaction";

import {
  sendAndConfirmFactory,
  setup,
} from "../src/lib.js";

const DEPOSIT_AMOUNT = 1_000_000_000n;
const TRANSFER_AMOUNT = 300_000_000n;
const WITHDRAW_AMOUNT = 300_000_000n;

async function main(): Promise<void> {
  const {
    sender: senderKeypair,
    recipient: recipientKeypair,
    clientConfig,
  } = await setup();

  // Connect to Helius devnet RPC plus the Photon indexer and prover.
  const client =
    await createZolanaClient(clientConfig);

  // Initialize the sender's private wallet and local authority
  // to decrypt transactions and sync balances.
  // The Solana signer and private wallet are derived from the same Ed25519 seed.
  const senderSigner =
    senderKeypair.toSolanaSigner();
  const senderAddress =
    senderKeypair.shieldedAddress();
  const recipient =
    recipientKeypair.shieldedAddress();

  // The SDK hands back instructions; the app owns signing and sending.
  const sendAndConfirm = sendAndConfirmFactory(
    client,
    senderSigner,
  );

  // Mints that are registered with Solana Rings for privacy.
  const assets = new AssetRegistry();

  // Deposit SOL into the sender's private balance.
  // A deposit from a public balance reveals
  // sender, recipient, asset and amount.
  // Alternatively, you can onramp fiat directly to a private balance.

  // 1. Move public SOL into the sender's private balance.
  // The view tag is the sender's Solana public key in confidential rings.
  // Used by the indexer to fetch the sender's outputs.
  const senderViewTag =
    senderAddress.confidentialViewTag();
  const depositIx = await depositInstruction({
    tree: client.tree,
    depositor: senderSigner,
    deposits: [
      {
        asset: DepositAsset.sol(),
        viewTag: senderViewTag,
        recipientOwnerHash:
          senderAddress.ownerHash(),
        blinding: randomBlinding(),
        amount: DEPOSIT_AMOUNT,
      },
    ],
  });

  // 2. Send and confirm like any Solana transaction; confirmation yields the landed slot.
  const depositTx = await sendAndConfirm([
    depositIx,
  ]);

  // 3. Fetch transaction outputs from the indexer, gated on the deposit's slot.
  // The indexer returns encrypted outputs by view tag.
  const depositResponse =
    await client.getShieldedTransactionsByTags(
      { tags: [senderViewTag] },
      atSlot(depositTx.slot),
    );

  // 4. The sender decrypts the transaction outputs locally to read the private balance.
  const balancesAfterDeposit =
    await decryptToBalances({
      keypair: senderKeypair,
      registry: assets,
      transactions: depositResponse.transactions,
    });
  const depositBalance =
    balancesAfterDeposit.balance(SOL_MINT);
  if (depositBalance.amount !== DEPOSIT_AMOUNT) {
    throw new Error(
      `expected deposit amount ${DEPOSIT_AMOUNT}, got ${depositBalance.amount}`,
    );
  }
  if (depositBalance.utxos.length !== 1) {
    throw new Error(
      `expected 1 deposit utxo, got ${depositBalance.utxos.length}`,
    );
  }

  // Confidential SOL transfer to the recipient's private balance.
  // A confidential transfer reveals only sender and recipient,
  // not the asset or amount.

  // 1. Select private token accounts (UTXOs) that make up the private balance for the transfer.
  const transferUtxo = depositBalance.utxos[0]!;

  // 2. Prepare the selected UTXOs as inputs for the zero-knowledge proof.
  const transferInput =
    ProofInputUtxo.fromKeypair(
      transferUtxo,
      senderKeypair,
    );

  // 3. Build and sign the confidential transfer.
  // Signing encrypts the asset and amount and produces the proof inputs for the ZK prover.
  const transfer = new ConfidentialTransfer(
    senderAddress,
    [transferInput],
    senderSigner.address,
  );
  transfer.send(
    recipient,
    SOL_MINT,
    TRANSFER_AMOUNT,
  );
  const transferProofInputs = transfer.sign(
    senderKeypair,
    assets,
  );

  // 4. Fetch the ZK proof to prove the sender can spend the balance without revealing asset and amount.
  const transferData = await client.proveTransact(
    transferProofInputs,
  );

  // 5. Build the instruction with the state Merkle tree and Solana accounts required for the transfer.
  // Private transfers move balances only between private token accounts, not public token accounts.
  const transferInstruction = transactInstruction(
    {
      payer: senderSigner,
      inputTree: client.tree,
      outputTree: client.tree,
      data: transferData,
    },
  );

  // 6. Send and confirm like any Solana transaction; confirmation yields the landed slot.
  const transferTx = await sendAndConfirm([
    transferInstruction,
  ]);

  // 7. Fetch the sender's outputs again, gated on the transfer's slot,
  // and read the remaining private balance.
  const transferResponse =
    await client.getShieldedTransactionsByTags(
      { tags: [senderViewTag] },
      atSlot(transferTx.slot),
    );
  const balancesAfterTransfer =
    await decryptToBalances({
      keypair: senderKeypair,
      registry: assets,
      transactions: transferResponse.transactions,
    });
  const transferBalance =
    balancesAfterTransfer.balance(SOL_MINT);
  if (
    transferBalance.amount !==
    DEPOSIT_AMOUNT - TRANSFER_AMOUNT
  ) {
    throw new Error(
      `expected remaining amount ${DEPOSIT_AMOUNT - TRANSFER_AMOUNT}, got ${transferBalance.amount}`,
    );
  }
  if (transferBalance.utxos.length !== 1) {
    throw new Error(
      `expected 1 transfer utxo, got ${transferBalance.utxos.length}`,
    );
  }

  // Withdraw SOL from the sender's private balance to their public balance.
  // A withdrawal reveals the sender, recipient, asset, and amount.

  // 1. Select private token accounts (UTXOs) that make up the private balance for the withdrawal.
  const withdrawalUtxo =
    transferBalance.utxos[0]!;

  // 2. Prepare the selected UTXOs as inputs for the zero-knowledge proof.
  const withdrawalInput =
    ProofInputUtxo.fromKeypair(
      withdrawalUtxo,
      senderKeypair,
    );

  // 3. Build and sign the private-to-public withdrawal.
  // Signing encrypts the asset and amount of the remaining private balance
  // and produces the proof inputs for the ZK prover.
  const withdrawal = new ConfidentialTransfer(
    senderAddress,
    [withdrawalInput],
    senderSigner.address,
  );
  withdrawal.withdraw(
    SOL_MINT,
    WITHDRAW_AMOUNT,
    WithdrawalTarget.sol({
      recipient: senderSigner.address,
    }),
  );
  const withdrawalProofInputs = withdrawal.sign(
    senderKeypair,
    assets,
  );

  // 4. Fetch the ZK proof to prove the sender can spend the balance.
  const withdrawalData =
    await client.proveTransact(
      withdrawalProofInputs,
    );

  // 5. Build the instruction with the state Merkle tree and Solana accounts required for the withdrawal.
  const withdrawalInstruction =
    transactInstruction({
      payer: senderSigner,
      inputTree: client.tree,
      outputTree: client.tree,
      withdrawal: TransactWithdrawal.sol({
        recipient: senderSigner.address,
      }),
      data: withdrawalData,
    });

  // 6. Send and confirm like any Solana transaction; confirmation yields the landed slot.
  const withdrawalTx = await sendAndConfirm([
    withdrawalInstruction,
  ]);

  // 7. Fetch the sender's outputs again, gated on the withdrawal's slot,
  // and read the remaining private balance.
  const withdrawalResponse =
    await client.getShieldedTransactionsByTags(
      { tags: [senderViewTag] },
      atSlot(withdrawalTx.slot),
    );
  const balancesAfterWithdrawal =
    await decryptToBalances({
      keypair: senderKeypair,
      registry: assets,
      transactions:
        withdrawalResponse.transactions,
    });
  const withdrawalBalance =
    balancesAfterWithdrawal.balance(SOL_MINT);
  if (
    withdrawalBalance.amount !==
    DEPOSIT_AMOUNT -
      TRANSFER_AMOUNT -
      WITHDRAW_AMOUNT
  ) {
    throw new Error(
      `expected remaining amount ${DEPOSIT_AMOUNT - TRANSFER_AMOUNT - WITHDRAW_AMOUNT}, got ${withdrawalBalance.amount}`,
    );
  }
  if (withdrawalBalance.utxos.length !== 1) {
    throw new Error(
      `expected 1 withdrawal utxo, got ${withdrawalBalance.utxos.length}`,
    );
  }

  // 8. Read remaining private balance and the public balance.
  const solanaBalance = await client.getBalance(
    senderSigner.address,
  );
  console.log(
    `withdraw private_balance=${withdrawalBalance.amount} ` +
      `solana_balance=${solanaBalance} tx=${withdrawalTx.signature}`,
  );
}

await main();
